Src/GoogleApis.Core/Apis/Http/ConfigurableMessageHandler.cs

```diff
@@ -64,22 +74,115 @@ public class ConfigurableMessageHandler : DelegatingHandler
         private readonly IList<IHttpExecuteInterceptor> executeInterceptors =
             new List<IHttpExecuteInterceptor>();
 
-        /// <summary>Gets a list of <see cref="IHttpUnsuccessfulResponseHandler"/>.</summary>
-        public IList<IHttpUnsuccessfulResponseHandler> UnsuccessfulResponseHandlers
+        /// <summary>
+        /// Gets a copied list of <see cref="IHttpUnsuccessfulResponseHandler"/>.
+        /// <remarks>
+        /// Since version 1.10, the return value had been changed from <c>IList</c> to <c>IEnumerable</c> in order to
+        /// keep this class thread-safe. The returned list is also a copied version of the handlers. More info is
+        /// available on <a href="https://github.com/google/google-api-dotnet-client/issues/592">Issue 592</a>.
+        /// </remarks>
+        /// </summary>
+        public IEnumerable<IHttpUnsuccessfulResponseHandler> UnsuccessfulResponseHandlers
+        {
+            get
+            {
+                lock (unsuccessfulResponseHandlersLock)

```

Comment:
A common practice for thread safe class is to have 1 lock object for the entire class and it usually works pretty well and makes things simple.
About affecting performance: I don't think performance is important in this case API and also I would also be very very surprised if the performance in a real life scenario was affected by having one lock object instead of 3. So my opinion is you are trying to optimize prematurely and also introducing unneccessary boilerplate.

I don't feel too strongly about this though, so if you are 100% convinced what you are doing is right, I'd be fine with that.


Regarding the comment: I'd change to "Gets a snapshot of unsuccessful response handlers list associated with this instance." or something in that sense.